Simplify s3 event with custom bucket configuration

richarddd opened the pull request:
## What did you implement

S3 events can now point to custom CF resources using the `resourceId` property:

```yaml
functions:
  resize:
    handler: hello.handler
    events:
      - s3: 
          resouceId: MyAwesomeS3Bucket

resources:
  Resources:
    MyAwesomeS3Bucket:
      Type: AWS::S3::Bucket
      Properties:
        BucketName: my-custom-bucket-name
        # add additional custom bucket configuration here
```

This removes a lot of boilerplate and code repetition when wanting to customize the bucket and use it as an event source.

## How does it work

If the `resourceId` property exists we will check if there is an existing resource with that logical id. If so, the bucket name is fetched from that resource and used rather than the `bucketName` of the event property. The logical id is also used in the new logical id of permission resources to keep it consistent with other resources.

## How can we verify it

Create a new project with the configuration above.

## Todos

- [x] Write and run all tests
- [x] Write documentation
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

**_Is this ready for review?:_** YES
**_Is it a breaking change?:_** NO

medikoo commented on lib/plugins/aws/package/compile/events/s3/index.js:
```diff
@@ -34,17 +38,20 @@ class AwsCompileS3Events {
-                  `Missing "bucket" property for s3 event in function ${functionName}.`,
-                  ' The correct syntax is: s3: bucketName OR an object with "bucket" property.',
+                  `Missing "${(!event.s3.bucket && 'bucket') ||
+                    'id'}" property for s3 event in function ${functionName}.`,
```
I believe it should be  fixed to `Missing "bucket" or "resoureId" property..`

medikoo commented on lib/plugins/aws/package/compile/events/s3/index.js:
```diff
@@ -21,7 +21,11 @@ class AwsCompileS3Events {
     const bucketsLambdaConfigurations = {};
+    const bucketsLogicalIds = {};
     const s3EnabledFunctions = [];
+    const resources =
```
Let's name it `userResources` (to not confuse with resource data that's generated by the framework)

medikoo commented on lib/plugins/aws/package/compile/events/s3/index.js:
```diff
@@ -143,13 +166,14 @@ class AwsCompileS3Events {
         const lambdaPermissionLogicalId = this.provider.naming.getLambdaS3PermissionLogicalId(
           item.functionName,
-          item.bucketName
+          bucketsLogicalIds[item.bucketName] || item.bucketName
```
I think we don't need to do any change here

richarddd replied:
The reason why I kept it this way is that I don't want the generated logical id to contain the bucket name. Assume you want to override the generated permission and have a custom bucket name containing the current `stage` ie: `my-custom-bucket-dev`. To override the generated Permission you have to specify this resource:
```
HelloLambdaPermissionMycustombucketdevS3:
      Type: 'AWS::Lambda::Permission'
      Properties:
```

But then it becomes problematic if you deploy a different stage as you'll have a different generated logical id

My point is that if you do custom s3 bucket configuration you want to avoid having dynamic bucket name in the logical id :)

medikoo replied:
Ok, so as user, when defining bucketName on his own, can make it dynamic (e.g. adapt to current stage, region etc.). It's more convinient to align to resource id, which shouldn't change across those.

Sounds reasonable, thanks for explanation

medikoo commented on lib/plugins/aws/package/compile/events/s3/index.js:
```diff
@@ -184,7 +208,7 @@ class AwsCompileS3Events {
       const lambdaPermissionLogicalId = this.provider.naming.getLambdaS3PermissionLogicalId(
         s3EnabledFunction.functionName,
-        s3EnabledFunction.bucketName
+        bucketsLogicalIds[s3EnabledFunction.bucketName] || s3EnabledFunction.bucketName
```
Same here

medikoo commented on lib/plugins/aws/package/compile/events/s3/index.js:
```diff
@@ -34,17 +38,19 @@ class AwsCompileS3Events {
                 const errorMessage = [
-                  `Missing "bucket" property for s3 event in function ${functionName}.`,
-                  ' The correct syntax is: s3: bucketName OR an object with "bucket" property.',
+                  `Missing "bucket" or "resoureId" property for s3 event in function ${functionName}.`,
```
`resourceId`

medikoo commented on lib/plugins/aws/package/compile/events/s3/index.js:
```diff
@@ -34,17 +38,19 @@ class AwsCompileS3Events {
 
             if (typeof event.s3 === 'object') {
-              if (!event.s3.bucket) {
+              if (!event.s3.bucket && !event.s3.resoureId) {
```
`resourceId`

medikoo commented on lib/plugins/aws/package/compile/events/s3/index.js:
```diff
@@ -78,17 +84,33 @@ class AwsCompileS3Events {
               const errorMessage = [
                 `S3 event of function ${functionName} is not an object nor a string.`,
-                ' The correct syntax is: s3: bucketName OR an object with "bucket" property.',
+                ' The correct syntax is: s3: bucketName OR an object with "bucket" or "id" property.',
```
`resourceId`

medikoo replied:
Doesn't seem to be addressed

medikoo commented on lib/plugins/aws/package/compile/events/s3/index.js:
```diff
@@ -78,17 +84,33 @@ class AwsCompileS3Events {
+            if (logicalId) {
+              if (!userResources[logicalId]) {
+                const errorMessage = [
+                  `S3 event has id property "${logicalId}" but no resource with that logical id could be found.`,
```
"has resourceId property"

medikoo commented on lib/plugins/aws/package/compile/events/s3/index.js:
```diff
@@ -78,17 +84,33 @@ class AwsCompileS3Events {
+              if (!userResources[logicalId]) {
+                const errorMessage = [
+                  `S3 event has id property "${logicalId}" but no resource with that logical id could be found.`,
+                  ` Please check that you have a S3 bucket in your "Resources:" with logcal id "${logicalId}"`,
```
_logical_

medikoo commented on lib/plugins/aws/package/compile/events/s3/index.test.js:
```diff
@@ -51,7 +51,7 @@ describe('AwsCompileS3Events', () => {
     });
 
-    it('should throw an error if the "bucket" property is not given', () => {
+    it('should throw an error if the "bucket" or "id" property is not given', () => {
```
`resourceId`

medikoo replied:
Also here

richarddd commented:
> Looks good, just spotted a few typos

Lol you misspelled it in your first comment and i was too lazy to write `resources` all over the place so copied it everywhere and introduced typos all over the place 😆

Will fix right away!

richarddd commented:
> Thank you @richarddd, we're near there :) There's just two more spots where "id" is put instead of "resourceId". I commented them out

Doh! Great catch! They are now removed and (hopefully) ready to be merged :)

richarddd commented:
> Hey @richarddd thanks for working on this 
> 
> This looks good from a code perspective. While I see the benefit I'm a little bit concerned about the contract / abstraction we're trying to establish here. There's #3448 which proposes to implement the missing `s3` event configs.
> 
> I personally prefer to add the config options rather than having a config to point to CloudFormation.
> 
> If we go down that route we should add this option to all the other event sources as well. Otherwise it'll be rather inconsistent.

Agree. I think we should add the `resouceId` ability to all other event sources as well. Adding more configuration to the event property of the function will create issues if we were to use the same event source for multiple functions. IMO it's strange that one event specification should be "master" and other consumers. Consider this use case:

```
function1:
  event:
    - s3:
        bucket: s3bucketname
        acessControll:...
        cors:...
        event: s3:ObjectRemoved:*
function2:
  event:
    - s3:
        bucket: s3bucketname
        event: s3:ObjectCreated:*
```

Of if we use the same bucket for multiple events on the same function:

```

function1:
  event:
    - s3:
        bucket: s3bucketname
        acessControll:...
        cors:...
        event: s3:ObjectRemoved:*
    - s3:
        bucket: s3bucketname
        event: s3:ObjectCreated:*
```

It can become problematic to know which function generated the "advanced" configuration.
The same principle applies to all other event sources as well.

This will be a bigger PR and maybe we could do it in steps. There are some shared code here which could be refactored out in a utils file such as checking if the `resourceId` exists in `userResources` and that the type matches the event type.

medikoo commented:
In case where we want to create an S3 bucket that triggers multiple lambdas, I think it's definitely better to not put general S3 configuration under one lambda.

e.g. in case of API Gateway it's sorted in a way that we put general API Gateway config into `provider.apiGateway`. Still we do not support more than one apiGateway that way.
I guess in case of S3 buckets we should be able to support multiple ones so e.g. as:

```yaml
provider:
  s3:
    photos:
      prop: ...
    videos:
      prop: ...
```

And adding support for such, would be probably best solution for this case, and #3448

richarddd commented:
> In case where we want to create an S3 bucket that triggers multiple lambdas, I think it's definitely better to not put general S3 configuration under one lambda.
> 
> e.g. in case of API Gateway it's sorted in a way that we put general API Gateway config into `provider.apiGateway`. Still we do not support more than one apiGateway that way.
> I guess in case of S3 buckets we should be able to support multiple ones so e.g. as:
> 
> ```yaml
> provider:
>   s3:
>     photos:
>       prop: ...
>     videos:
>       prop: ...
> ```
> 
> And adding support for such, would be probably best solution for this case, and #3448

Well, it then introduces a lot of repetition when we use stage (or other) variables in our bucket name as we would need to specify those everywhere. That could be partially solved by using a custom value but still a bit tedius to use something like: `${self: custom.myBucketName}` in several places.

My point is that using dynamic naming as a key (reference if you will) is not so ideal.

medikoo commented:
> Well, it then introduces a lot of repetition when we use stage (or other) variables in our bucket name 

Technically we could also support format as:

```yaml
provider:
  s3:
    - name: photos
      prop: ...
    - name: videos
      prop: ...
```

Or implement support for vars resolution in property names (doable, but probably more challenging)

richarddd commented:
> ```yaml
> photos
> ```

And how would you reference it then in your event? It would be the same problem, right?

What I'm trying to avoid is something like this:

```
custom:
   myBucketNameOne: ${self:provider.stage}-my-bucket-one
   myBucketNameTwo: ${self:provider.stage}-my-bucket-two

provider:
  s3:
    - name: ${self:custom.myBucketNameOne}
      prop: ...
    - name: ${self:custom.myBucketNameTwo}
      prop: ...

functions:
   functionOne:
      events:
         - s3:
               bucket: ${self:custom.myBucketNameOne}
               event: s3:ObjectCreated:*
         - s3:
               bucket: ${self:custom.myBucketNameOne}
               event: s3: ObjectRemoved:*
   functionTwo:
      events:
         - s3:
               bucket: ${self:custom.myBucketNameTwo}
               event: s3:ObjectCreated:*
         - s3:
               bucket: ${self:custom.myBucketNameTwo}
               event: s3: ObjectRemoved:*
```

If we could do something like:
```
provider:
  s3:
    - referenceId: bucketOne 
      name: ${self:provider.stage}-my-bucket-one
      prop: ...
     - referenceId: bucketTwo
      name: ${self:provider.stage}-my-bucket-two
      prop: ...

functions:
   functionOne:
      events:
         - s3:
               referenceId: bucketOne
               event: s3:ObjectCreated:*
```

It would work as well

medikoo commented:
> If we could do something like (...) It would work as well

That's what I had in mind in last comment

richarddd commented:
> > If we could do something like (...) It would work as well
> 
> That's what I had in mind in last comment

Cool! Will implement this instead then! What should be call the `id` prop then? `name` could be confused with `bucketName` IMO

medikoo commented:
@richarddd Sorry I quickly replied before, without fully understanding your point.

I think we can do better, and follow configuration method that's already used for naming functions. So:

```yaml
provider:
  s3:
    bucketOne:
      # name is optional, if not provided it's `bucketOne` that's picked for bucket name
      name: ${self:provider.stage}-my-bucket-one
      prop: ...
  functions:
    function:
      events:
        - s3:
          bucket: bucketOne
```

@pmuens what do you think?

richarddd commented:
> @richarddd Sorry I quickly replied before, without fully understanding your point.
> 
> I think we can do better, and follow configuration method that's already used for naming functions. So:
> 
> ```yaml
> provider:
>   s3:
>     bucketOne:
>       # name is optional, if not provided it's `bucketOne` that's picked for bucket name
>       name: ${self:provider.stage}-my-bucket-one
>       prop: ...
>   functions:
>     function:
>       events:
>         - s3:
>           bucket: bucketOne
> ```
> 
> @pmuens what do you think?

@medikoo @pmuens what do you guys say about the above? It would work perfect IMO :)

pmuens commented:
I'm still not sold on the idea of referencing raw CloudFormation within an abstraction.

However putting the logic under provider makes sense given the discussion both of you had.

medikoo commented:
> I'm still not sold on the idea of referencing raw CloudFormation within an abstraction.

Yes, we definitely revert from that

> However putting the logic under provider makes sense given the discussion both of you had.

That's the plan.

@richarddd therefore we're all after having it as proposed in my last comment

richarddd commented:
> > I'm still not sold on the idea of referencing raw CloudFormation within an abstraction.
> 
> Yes, we definitely revert from that
> 
> > However putting the logic under provider makes sense given the discussion both of you had.
> 
> That's the plan.
> 
> @richarddd therefore we're all after having it as proposed in my last comment

@medikoo here we go :)

1. The latest commit has the specification we discussed above. The generated logical id for buckets with a custom `provider.s3` entry will be based on the "key" of that entry. The reason for this is so we don't have the bucket name (which is likely dynamic) as a basis for our logical id generation. The generated logical id will still follow the naming standard. Again, this ONLY applies when the `provider.s3` has a match with the `event.s3` value.

2. Furthermore, I also added a check on both property key and type according to the CloudFormation spec of buckets so that invalid properties are caught as early as possible.

3. NotificationConfiguration is now merged with the custom properties.

4. Added a bunch of tests to cover these cases

medikoo commented on lib/plugins/aws/package/compile/events/s3/index.js:
```diff
@@ -21,7 +45,10 @@ class AwsCompileS3Events {
 
   newS3Buckets() {
     const bucketsLambdaConfigurations = {};
+    const bucketsLogicalIds = {};
```
Name suggests it's a logicalIds map, while it's something more, let's name it then maybe `bucketsMeta`

medikoo replied:
It'll be nice to use `Map` here. Note that such plain object notation is not safe for hacky bucket names as `hasOwnProperty`

medikoo commented on lib/plugins/aws/package/compile/events/s3/index.js:
```diff
@@ -74,21 +102,32 @@ class AwsCompileS3Events {
+            let bucketName;
+            if (provderS3[bucket]) {
+              bucketName = provderS3[bucket].name || bucket;
+              logicalId = this.provider.naming.getBucketLogicalId(bucket);
```
`logicalId` could be defined purely in scope of this block, and here we can also create a _bucket meta_ object

medikoo commented on lib/plugins/aws/package/compile/events/s3/index.js:
```diff
@@ -126,14 +165,59 @@ class AwsCompileS3Events {
     _.forEach(bucketsLambdaConfigurations, (bucketLambdaConfiguration, bucketName) => {
+      let bucketProps = null;
+      if (bucketsLogicalIds[bucketName]) {
+        const providerBucket = provderS3[bucketsLogicalIds[bucketName].bucket];
```
Let's name this var `bucketConf`

medikoo commented on lib/plugins/aws/package/compile/events/s3/index.js:
```diff
@@ -126,14 +165,59 @@ class AwsCompileS3Events {
+      let bucketProps = null;
+      if (bucketsLogicalIds[bucketName]) {
+        const providerBucket = provderS3[bucketsLogicalIds[bucketName].bucket];
+        if (providerBucket) {
```
Logically it'll always be true, therefore this condition seems redundant

medikoo commented on lib/plugins/aws/package/compile/events/s3/index.js:
```diff
@@ -126,14 +165,59 @@ class AwsCompileS3Events {
+      if (bucketsLogicalIds[bucketName]) {
+        const providerBucket = provderS3[bucketsLogicalIds[bucketName].bucket];
+        if (providerBucket) {
+          bucketProps = Object.keys(providerBucket).reduce((acc, key) => {
```
`reduce` is a controversial and confusuing choice for such operation. Can we just iterate over entries as:

```javascript
bucketProps = {};
for (const [key, name] of _.entries(bucketConf)) {
  ...
}
```

medikoo commented on lib/plugins/aws/package/compile/events/s3/index.js:
```diff
@@ -126,14 +165,59 @@ class AwsCompileS3Events {
+                  'A bucket could only be configured with the following properties:\n',
+                  ['name'].concat(_.keys(this.allowedBucketProperties)).join(', '),
+                ].join('');
+                throw new this.serverless.classes.Error(errorMessage);
```
I would not introduce such validation here, as we should have such validation solved globally for whole configuration (there's a dedicated issue: https://github.com/serverless/serverless/issues/6562 )

So maybe just accept recognized properties, ignore others, and do not do type checking (it's covered already by CF validation which is done prior deployment)

medikoo commented on lib/plugins/aws/package/compile/events/s3/index.js:
```diff
@@ -74,21 +102,32 @@ class AwsCompileS3Events {
+            }
+
+            if (logicalId) {
+              bucketsLogicalIds[bucketName] = { logicalId, bucket };
```
Let's put _name_ as `name` (not `bucket`)

richarddd replied:
Thats not always the name

medikoo replied:
So if it's our reference name, let's name it `ref`

richarddd commented:
@medikoo alrighty! should be updated according to your suggestions now

richarddd commented on lib/plugins/aws/package/compile/events/s3/index.js:
```diff
@@ -10,6 +10,27 @@ class AwsCompileS3Events {
 
+    this.allowedBucketProperties = _.keyBy([
+      'AccelerateConfiguration',
+      'accessControl',
```
Great catch!

richarddd commented:
> Looks great! Added just few minor suggestions
> 
> Also do you think we can maybe cover that with integration test that covers some properties (those straightforward) here: https://github.com/serverless/serverless/blob/0549d85bc0254a10d3314613892e335da2bc3722/tests/integration-all/s3/tests.js ?

Done! While doing so I found a bug with my implementation where bucket names could contain illegal chars. This is now patched and I created a test for it as well.

medikoo commented on lib/plugins/aws/package/compile/events/s3/index.js:
```diff
@@ -126,14 +173,46 @@ class AwsCompileS3Events {
+        for (const [key, value] of _.entries(providerBucket)) {
+          if (key !== 'name') {
+            const property = _.upperFirst(key);
+            const allowedProperety = this.allowedBucketProperties[property];
```
Typo (should be `allowedProperty`)

richarddd commented on lib/plugins/aws/package/compile/events/s3/index.js:
```diff
@@ -74,21 +101,41 @@ class AwsCompileS3Events {
+                this.provider.naming
+                  .normalizeBucketName(bucketRef)
+                  .replace(/\B([A-Z])/, '-$1')
+                  .toLowerCase();
```
The reason why this is used is so we can ensure bucket names are valid when we generate them. This is only used when we use an empty object in `provider.s3.myBucket`. If i dont add this, the resulting bucking name would be `myBucket` which is an invalid name. This like would instead create a bucket with the name `my-bucket`

medikoo replied:
> This is only used when we use an empty object in `provider.s3.myBucket`

Isn't it used when there's no `name` provided to  provider.s3.myBucket (so also when there are some other properties)?

Technically I think it'll be good to maintain similar convention as with function names (although function names allow capital letters, and framework also prefixes them with service and stage name).

Do you think it's an issue for a developer to ensure that used ref name is constructed only from characters that are allowed in bucket name?

richarddd replied:
> Isn't it used when there's no `name` provided to provider.s3.myBucket (so also when there are some other properties)?

Yes.

> Technically I think it'll be good to maintain similar convention as with function names (although function names allow capital letters, and framework also prefixes them with service and stage name).

Should we add the same logic here (stage + service name)? I think we must at least lower case them since CF will error otherwise by just following yaml key naming convention (camel case).

> Do you think it's an issue for a developer to ensure that used ref name is constructed only from characters that are allowed in bucket name?

Yes.

medikoo replied:
> Should we add the same logic here (stage + service name)? 

I wouldn't do it, as we don't do it with `bucketName` as put at `s3` event. Technically framework is already inconsistent, and trying to fix that will make a breaking change

>  I think we must at least lower case them since CF will error otherwise by just following yaml key naming convention (camel case).

The problem I see, is that if we mangle that name, then it becomes confusing. User cannot be sure what the final name of created bucket will be and how to find it.

AFAIK we don't do such mangling in other cases (aside of prefixing with stage and service name), so I'm not sure about introducing here.
I think we do such normalization in case of CF resource naming, but that's different thing (those ids are used only in CF template context)

@pmuens what do you think?

richarddd replied:
@medikoo what's your suggestion then? Lowercase it only? :)

medikoo replied:
I would pass as is 1:1, and let AWS template validation fail if invalid name is used.

Still let's wait for feedback from @pmuens

richarddd replied:
> I would pass as is 1:1, and let AWS template validation fail if invalid name is used.
> 
> Still let's wait for feedback from @pmuens

I think you're missing the point here. The deploy will probably rarely succeed since keys in `provider.s3` are very likely to be camel-cased. This modification does only apply when no name or bucketName property exists in the `provider.s3.myBucket` object.

**The following will succeed:**
```yml
provider:
  s3:
     myBucket:
       name: my-awesome-custom-bucket
```

**This will not (with your suggestion)**
```yml
provider:
  s3:
     myBucket:
       objectLockEnabled: true
```
_(☝️ the bucket name will be "myBucket" without the modifications I have made, which is invalid since it contains uppercase chars)_

medikoo replied:
@richarddd Thanks for explanations, Still I have full understanding of that, and my feeling is that it's better if user is forced to configure bucket names 1:1 as follows:

```yaml
provider:
  s3:
    myBucket:
      name: my-awesome-custom-bucket
      objectLockEnabled: true
```

Or

```yaml
provider:
  s3:
    my-awesome-custom-bucket:
      objectLockEnabled: true
```

And have full control and understanding of what final bucket name will be (no magical bucket name correction)

Being at stake of normalization algorithm (which while will allow him to put any string as bucket reference), will in many cases create a bucket with name he didn't put (I'm worried in many cases it'll turn confusing and surprising).

--- 

Note that Framework currently is not doing that. If someone will put:

```yaml
functions:
  function:
    handler: index.handler
    events:
      - s3: myBucket
```

Deployment will crash with ` An error occurred: S3BucketTestFoo - Bucket name should not contain uppercase characters.`

And I think it's better than implying normalization.

If we agree on what you're proposing, then just adding following to above config

```yaml
provider:
    s3:
      myBucket: {}
```

Will change that behavior, and create S3 with `mybucket` name. That doesn't feel right to me.

--- 

I understand that your argument is to allow users to use camelCase (natural) notation in YAML for bucket configuration, but I think  predictability of the outcome is more valuable here, then maintaining config file property names convention.

medikoo replied:
Other potential issue with normalization implied is that different refs may translate to same bucketName (e.g. `myBucket`, `mybucket`, `MY_BUCKET` etc will resolve to `mybucket`) causing collisions, which are probably not handled well.

richarddd replied:
@medikoo you have a valid point there! I came to think of swagger v3 spec where the routes/paths are indeed a 1:1 which feels quite natural. I'll update the PR and remove the normalization :)

richarddd replied:
btw, do you think we should remove this check also? I feel like maybe this would be great to catch in an early stage:  
https://github.com/serverless/serverless/pull/7156/files#diff-a5a4c36cd72d14b90ae83a1ed4bb1fdbR123

medikoo replied:
Good question, I agree it'll be convenient to abort early in invalid name case. I would  just move it out of condition, so it's validated in all configuration cases

richarddd replied:
Thanks for the feedback. Fixed in the latest push!

medikoo commented on lib/plugins/aws/package/compile/events/s3/index.js:
```diff
@@ -10,18 +10,45 @@ class AwsCompileS3Events {
     };
+
+    this.errorMessageBucketSyntax =
+      'The correct syntax is: "s3: bucketName or bucketRef or an object with "bucket" property.';
```
Let's make it a normal var in body of a module

medikoo commented on lib/plugins/aws/package/compile/events/s3/index.js:
```diff
@@ -74,21 +101,37 @@ class AwsCompileS3Events {
+              const errorMessage = [
+                `${bucketName} - Bucket name should not contain uppercase characters`,
+                `Please check provider.s3.${bucketRef} and/or s3 events of function "${functionName}"`,
+              ].join(' ');
```
If it's multisentence, it'll be good to end sentences with dot

richarddd commented:
@medikoo we should be good to go now :)
